File: linuxsdl/lesson09/lesson09.c
/*
 * This code was created by Jeff Molofee '99 
 * (ported to Linux/SDL by Ti Leggett '01)
 *
 * If you've found this code useful, please let me know.
 *
 * Visit Jeff at http://nehe.gamedev.net/
 * 
 * or for port-specific comments, questions, bugreports etc. 
 * email to leggett@eecs.tulane.edu
 */
 
#include <stdio.h>
#include <stdlib.h>
#include <GL/gl.h>
#include <GL/glu.h>
#include "SDL.h"

/* screen width, height, and bit depth */
#define SCREEN_WIDTH  640
#define SCREEN_HEIGHT 480
#define SCREEN_BPP     16

/* Set up some booleans */
#define TRUE  1
#define FALSE 0

/* Number of stars */
#define NUM 50

/* This is our SDL surface */
SDL_Surface *surface;

/* Twinkling stars */
int twinkle = FALSE;

/* Define the star structure */
typedef struct
{
    int r, g, b;   /* Stars Color */
    GLfloat dist;  /* Stars Distance From Center */
    GLfloat angle; /* Stars Current Angle */
} star;

star stars[NUM];       /* Make an array of size 'NUM' of stars */

GLfloat zoom = -15.0f; /* Viewing Distance Away From Stars */
GLfloat tilt = 90.0f;  /* Tilt The View */

GLuint loop;           /* General Loop Variable */
GLuint texture[1];     /* Storage For One Texture */


/* function to release/destroy our resources and restoring the old desktop */
void Quit( int returnCode )
{
    /* clean up the window */
    SDL_Quit( );

    /* and exit appropriately */
    exit( returnCode );
}

/* function to load in bitmap as a GL texture */
int LoadGLTextures( )
{
    /* Status indicator */
    int Status = FALSE;

    /* Create storage space for the texture */
    SDL_Surface *TextureImage[1]; 

    /* Load The Bitmap, Check For Errors, If Bitmap's Not Found Quit */
    if ( ( TextureImage[0] = SDL_LoadBMP( "data/star.bmp" ) ) )
        {

	    /* Set the status to true */
	    Status = TRUE;

	    /* Create The Texture */
	    glGenTextures( 1, &texture[0] );

	    /* Load in texture */
	    /* Typical Texture Generation Using Data From The Bitmap */
	    glBindTexture( GL_TEXTURE_2D, texture[0] );

	    /* Linear Filtering */
	    glTexParameteri( GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR );
	    glTexParameteri( GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR );

	    /* Generate The Texture */
	    glTexImage2D( GL_TEXTURE_2D, 0, 3, TextureImage[0]->w,
			  TextureImage[0]->h, 0, GL_BGR,
			  GL_UNSIGNED_BYTE, TextureImage[0]->pixels );

        }

    /* Free up any memory we may have used */
    if ( TextureImage[0] )
	    SDL_FreeSurface( TextureImage[0] );

    return Status;
}

/* function to reset our viewport after a window resize */
int resizeWindow( int width, int height )
{
    /* Height / width ration */
    GLfloat ratio;
 
    /* Protect against a divide by zero */
    if ( height == 0 )
	height = 1;

    ratio = ( GLfloat )width / ( GLfloat )height;

    /* Setup our viewport. */
    glViewport( 0, 0, ( GLint )width, ( GLint )height );

    /* change to the projection matrix and set our viewing volume. */
    glMatrixMode( GL_PROJECTION );
    glLoadIdentity( );

    /* Set our perspective */
    gluPerspective( 45.0f, ratio, 0.1f, 100.0f );

    /* Make sure we're chaning the model view and not the projection */
    glMatrixMode( GL_MODELVIEW );

    /* Reset The View */
    glLoadIdentity( );

    return( TRUE );
}

/* function to handle key press events */
void handleKeyPress( SDL_keysym *keysym )
{
    switch ( keysym->sym )
	{
	case SDLK_ESCAPE:
	    /* ESC key was pressed */
	    Quit( 0 );
	    break;
	case SDLK_F1:
	    /* F1 key was pressed
	     * this toggles fullscreen mode
	     */
	    SDL_WM_ToggleFullScreen( surface );
	    break;
	case SDLK_t:
	    /* 't' key was pressed
	     * this toggles the twinkling of the stars
	     */
	    twinkle = !twinkle;
	    break;
	case SDLK_UP:
	    /* Up arrow key was pressed
	     * this changes the tilt of the stars
	     */
	    tilt -= 0.5f;
	    break;
	case SDLK_DOWN:
	    /* Down arrow key was pressed
	     * this changes the tilt of the stars
	     */
	    tilt += 0.5f;
	    break;
	case SDLK_PAGEUP:
	    /* PageUp key was pressed
	     * zoom into the scene
	     */
	    zoom -= 0.2f;
	    break;
	case SDLK_PAGEDOWN:
	    /* PageDown key was pressed
	     * zoom out of the scene
	     */
	    zoom += 0.2f;
	    break;
	default:
	    break;
	}

    return;
}

/* general OpenGL initialization function */
int initGL( GLvoid )
{

    /* Load in the texture */
    if ( !LoadGLTextures( ) )
	return FALSE;

    /* Enable Texture Mapping */
    glEnable( GL_TEXTURE_2D );

    /* Enable smooth shading */
    glShadeModel( GL_SMOOTH );

    /* Set the background black */
    glClearColor( 0.0f, 0.0f, 0.0f, 0.0f );

    /* Depth buffer setup */
    glClearDepth( 1.0f );

    /* Really Nice Perspective Calculations */
    glHint( GL_PERSPECTIVE_CORRECTION_HINT, GL_NICEST );

    /* Set The Blending Function For Translucency */
    glBlendFunc( GL_SRC_ALPHA, GL_ONE );

    /* Enable Blending */
    glEnable( GL_BLEND );

    /* Create A Loop That Goes Through All The Stars */
    for ( loop = 0; loop < NUM; loop++ )
	{
	    /* Start All The Stars At Angle Zero */
	    stars[loop].angle = 0.0f;

	    /* Calculate Distance From The Center */
	    stars[loop].dist = ( ( float )loop / NUM ) * 5.0f;
	    /* Give star[loop] A Random Red Intensity */
	    stars[loop].r = rand( ) % 256;
	    /* Give star[loop] A Random Green Intensity */
	    stars[loop].g = rand( ) % 256;
	    /* Give star[loop] A Random Blue Intensity */
	    stars[loop].b = rand( ) % 256;
        }

    return( TRUE );
}

/* Here goes our drawing code */
int drawGLScene( GLvoid )
{
    static int spin = 0;

    /* These are to calculate our fps */
    static GLint T0     = 0;
    static GLint Frames = 0;

    /* Clear The Screen And The Depth Buffer */
    glClear( GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT );

    /* Select Our Texture */
    glBindTexture( GL_TEXTURE_2D, texture[0] );
    glLoadIdentity( );

    /* Loop Through All The Stars */
    for ( loop = 0; loop < NUM; loop++ )
        {
	    /* Reset The View Before We Draw Each Star */
	    glLoadIdentity( );
	    /* Zoom Into The Screen (Using The Value In 'zoom') */
	    glTranslatef( 0.0f, 0.0f, zoom );

	    /* Tilt The View (Using The Value In 'tilt') */
	    glRotatef( tilt, 1.0f, 0.0f, 0.0f );
	    /* Rotate To The Current Stars Angle */
	    glRotatef( stars[loop].angle, 0.0f, 1.0f, 0.0f );

	    /* Move Forward On The X Plane */
	    glTranslatef( stars[loop].dist, 0.0f, 0.0f );

	    /* Cancel The Current Stars Angle */
	    glRotatef( -stars[loop].angle, 0.0f, 1.0f, 0.0f );
	    /* Cancel The Screen Tilt */
	    glRotatef( -tilt, 1.0f, 0.0f, 0.0f );

	    /* Twinkling Stars Enabled */
	    if ( twinkle )
                {
		    /* Assign A Color Using Bytes */
		    glColor4ub( stars[( NUM - loop ) - 1].r,
				stars[( NUM - loop ) - 1].g,
				stars[( NUM - loop ) - 1].b, 255 );
		    /* Begin Drawing The Textured Quad */
		    glBegin( GL_QUADS );
		      glTexCoord2f( 0.0f, 0.0f );
		      glVertex3f( -1.0f, -1.0f, 0.0f );
		      glTexCoord2f( 1.0f, 0.0f);
		      glVertex3f( 1.0f, -1.0f, 0.0f );
		      glTexCoord2f( 1.0f, 1.0f );
		      glVertex3f( 1.0f, 1.0f, 0.0f );
		      glTexCoord2f( 0.0f, 1.0f );
		      glVertex3f( -1.0f, 1.0f, 0.0f );
		    glEnd( );
                }

	    /* Rotate The Star On The Z Axis */
	    glRotatef( spin, 0.0f, 0.0f, 1.0f );

	    /* Assign A Color Using Bytes */
	    glColor4ub( stars[loop].r, stars[loop].g, stars[loop].b, 255 );

	    /* Begin Drawing The Textured Quad */
	    glBegin( GL_QUADS );
	      glTexCoord2f( 0.0f, 0.0f ); glVertex3f( -1.0f, -1.0f, 0.0f );
	      glTexCoord2f( 1.0f, 0.0f ); glVertex3f(  1.0f, -1.0f, 0.0f );
	      glTexCoord2f( 1.0f, 1.0f ); glVertex3f(  1.0f,  1.0f, 0.0f );
	      glTexCoord2f( 0.0f, 1.0f ); glVertex3f( -1.0f,  1.0f, 0.0f );
	    glEnd( );

	    /* Used To Spin The Stars */
	    spin += 0.01f;
	    /* Changes The Angle Of A Star */
	    stars[loop].angle += ( float )loop / NUM;
	    /* Changes The Distance Of A Star */
	    stars[loop].dist -= 0.01f;

	    /* Is The Star In The Middle Yet */
	    if ( stars[loop].dist < 0.0f )
                {
		    /* Move The Star 5 Units From The Center */
		    stars[loop].dist += 5.0f;
		    /* Give It A New Red Value */
		    stars[loop].r = rand( ) % 256;
		    /* Give It A New Green Value */
		    stars[loop].g = rand( ) % 256;
		    /* Give It A New Blue Value */
		    stars[loop].b = rand( ) % 256;
                }
        }

    /* Draw it to the screen */
    SDL_GL_SwapBuffers( );

    /* Gather our frames per second */
    Frames++;
    {
	GLint t = SDL_GetTicks();
	if (t - T0 >= 5000) {
	    GLfloat seconds = (t - T0) / 1000.0;
	    GLfloat fps = Frames / seconds;
	    printf("%d frames in %g seconds = %g FPS\n", Frames, seconds, fps);
	    T0 = t;
	    Frames = 0;
	}
    }

    return( TRUE );
}

int main( int argc, char **argv )
{
    /* Flags to pass to SDL_SetVideoMode */
    int videoFlags;
    /* main loop variable */
    int done = FALSE;
    /* used to collect events */
    SDL_Event event;
    /* this holds some info about our display */
    const SDL_VideoInfo *videoInfo;
    /* whether or not the window is active */
    int isActive = TRUE;

    /* initialize SDL */
    if ( SDL_Init( SDL_INIT_VIDEO ) < 0 )
	{
	    fprintf( stderr, "Video initialization failed: %s\n",
		     SDL_GetError( ) );
	    Quit( 1 );
	}

    /* Fetch the video info */
    videoInfo = SDL_GetVideoInfo( );

    if ( !videoInfo )
	{
	    fprintf( stderr, "Video query failed: %s\n",
		     SDL_GetError( ) );
	    Quit( 1 );
	}

    /* the flags to pass to SDL_SetVideoMode */
    videoFlags  = SDL_OPENGL;          /* Enable OpenGL in SDL */
    videoFlags |= SDL_GL_DOUBLEBUFFER; /* Enable double buffering */
    videoFlags |= SDL_HWPALETTE;       /* Store the palette in hardware */
    videoFlags |= SDL_RESIZABLE;       /* Enable window resizing */

    /* This checks to see if surfaces can be stored in memory */
    if ( videoInfo->hw_available )
	videoFlags |= SDL_HWSURFACE;
    else
	videoFlags |= SDL_SWSURFACE;

    /* This checks if hardware blits can be done */
    if ( videoInfo->blit_hw )
	videoFlags |= SDL_HWACCEL;

    /* Sets up OpenGL double buffering */
    SDL_GL_SetAttribute( SDL_GL_DOUBLEBUFFER, 1 );

    /* get a SDL surface */
    surface = SDL_SetVideoMode( SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_BPP,
				videoFlags );

    /* Verify there is a surface */
    if ( !surface )
	{
	    fprintf( stderr,  "Video mode set failed: %s\n", SDL_GetError( ) );
	    Quit( 1 );
	}

    /* Enable key repeat */
    if ( ( SDL_EnableKeyRepeat( 100, SDL_DEFAULT_REPEAT_INTERVAL ) ) )
	{
	    fprintf( stderr, "Setting keyboard repeat failed: %s\n",
		     SDL_GetError( ) );
	    Quit( 1 );
	}

    /* initialize OpenGL */
    initGL( );

    /* resize the initial window */
    resizeWindow( SCREEN_WIDTH, SCREEN_HEIGHT );

    /* wait for events */
    while ( !done )
	{
	    /* handle the events in the queue */

	    while ( SDL_PollEvent( &event ) )
		{
		    switch( event.type )
			{
			case SDL_ACTIVEEVENT:
			    /* Something's happend with our focus
			     * If we lost focus or we are iconified, we
			     * shouldn't draw the screen
			     */
			    if ( event.active.gain == 0 )
				isActive = FALSE;
			    else
				isActive = TRUE;
			    break;			    
			case SDL_VIDEORESIZE:
			    /* handle resize event */
			    surface = SDL_SetVideoMode( event.resize.w,
							event.resize.h,
							16, videoFlags );
			    if ( !surface )
				{
				    fprintf( stderr, "Could not get a surface after resize: %s\n", SDL_GetError( ) );
				    Quit( 1 );
				}
			    resizeWindow( event.resize.w, event.resize.h );
			    break;
			case SDL_KEYDOWN:
			    /* handle key presses */
			    handleKeyPress( &event.key.keysym );
			    break;
			case SDL_QUIT:
			    /* handle quit requests */
			    done = TRUE;
			    break;
			default:
			    break;
			}
		}

	    /* draw the scene */
	    if ( isActive )
		drawGLScene( );
	}

    /* clean ourselves up and exit */
    Quit( 0 );

    /* Should never get here */
    return( 0 );
}

File: c_sharp/lesson09/lesson09.cs
/* 
 * NeHe Lesson 09 - Moving bitmaps in 3D space
 * http://nehe.gamedev.net/data/lessons/lesson.asp?lesson=09
 * 
 * Ported to C# by Brian Holley, February, 2004
 * http://tachyon.unl.edu
 * 
 * Uses the CsGL library from http://csgl.sourceforge.net
 * Some CsGL setup and image code from NeHe Lesson 06 C# port by Sabine Felsinger
 */

using System;
using System.Drawing;
using System.Windows.Forms;
using CsGL.OpenGL;

namespace Lesson09
{
	public class MainForm : System.Windows.Forms.Form	
	{
		public Lesson09.LessonView view;

		public MainForm()
		{
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(640, 480);
			this.Name = "MainForm";
			this.Text = "NeHe Lesson 09 - C#";
			this.view = new Lesson09.LessonView();
			this.view.Parent = this;
			this.view.Dock = DockStyle.Fill;
			this.Show();
		}

		static void Main() 
		{
			MainForm form = new MainForm();

			while ((!form.view.finished) && (!form.IsDisposed))
			{
				form.view.glDraw();
				form.Refresh();
				Application.DoEvents();
			}

			form.Dispose();
		}
	}

	// Start public class
	public class Star
	{
		public byte r, g, b;		// Star Color
		public float dist;			// Star Distance From Center
		public float angle;		// Star Current Angle

		public Star()
		{
			this.r = this.g = this.b = 0;
			this.dist = this.angle = 0.0f;
		}
	}
		
	public class LessonView : OpenGLControl
	{
		public bool twinkle = true;					// Twinkling Stars
		public bool tp = false;						// 'T' Key Pressed?

		public static int num = 50;					// Number Of Stars To Draw
		public Star[] star = null;					// Need To Keep Track Of 'num' Stars

		public float zoom = -15.0f;					// Distance Away From Stars
		public float tilt = 90.0f;					// Tilt The View
		public float spin = 0.0f;					// Spin Stars
		
		public Random rand = new Random();			// Random number generator

		public uint	loop = 0;						// General Loop Variable
		public uint[] texture = new uint[1];		// Storage For One Texture

		public bool finished;
        
		public LessonView() : base()
		{
			this.KeyDown += new KeyEventHandler(LessonView_KeyDown);
			this.KeyUp += new KeyEventHandler(LessonView_KeyUp);
			this.star = new Star[LessonView.num];
			for (int i=0; i < LessonView.num; i++)
				this.star[i] = new Star();
			this.finished = false;
		}

		protected override void InitGLContext() 
		{
			LoadTextures();

			GL.glEnable(GL.GL_TEXTURE_2D);									// Enable Texture Mapping
			GL.glShadeModel(GL.GL_SMOOTH);									// Enable Smooth Shading
			GL.glClearColor(0.0f, 0.0f, 0.0f, 0.5f);						// Black Background
			GL.glClearDepth(1.0f);											// Depth Buffer Setup
			GL.glHint(GL.GL_PERSPECTIVE_CORRECTION_HINT, GL.GL_NICEST);		// Really Nice Perspective Calculations
			GL.glBlendFunc(GL.GL_SRC_ALPHA, GL.GL_ONE);						// Set The Blending Function For Translucency
			GL.glEnable(GL.GL_BLEND);

			for (this.loop=0; this.loop < LessonView.num; this.loop++)
			{
				this.star[this.loop].angle = 0.0f;
				this.star[this.loop].dist = ((float)(this.loop) / LessonView.num) * 5.0f;
				this.star[this.loop].r = (byte)rand.Next(256);
				this.star[this.loop].g = (byte)rand.Next(256);
				this.star[this.loop].b = (byte)rand.Next(256);
			}
		}

		protected bool LoadTextures()
		{
			Bitmap image = null;
			string file = @"Data\Star.bmp";
			try
			{
				// If the file doesn't exist or can't be found, an ArgumentException is thrown instead of
				// just returning null
				image = new Bitmap(file);
			} 
			catch (System.ArgumentException)
			{
				MessageBox.Show("Could not load " + file + ".  Please make sure that Data is a subfolder from where the application is running.", "Error", MessageBoxButtons.OK);
				this.finished = true;
			}
			if (image != null)
			{
				image.RotateFlip(RotateFlipType.RotateNoneFlipY);
				System.Drawing.Imaging.BitmapData bitmapdata;
				Rectangle rect = new Rectangle(0, 0, image.Width, image.Height);

				bitmapdata = image.LockBits(rect, System.Drawing.Imaging.ImageLockMode.ReadOnly, System.Drawing.Imaging.PixelFormat.Format24bppRgb);

				GL.glGenTextures(1, this.texture);
			
				// Create Linear Filtered Texture
				GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[0]);
				GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MAG_FILTER, GL.GL_LINEAR);
				GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MIN_FILTER, GL.GL_LINEAR);
				GL.glTexImage2D(GL.GL_TEXTURE_2D, 0, (int)GL.GL_RGB, image.Width, image.Height, 0, GL.GL_BGR_EXT, GL.GL_UNSIGNED_BYTE, bitmapdata.Scan0);

				image.UnlockBits(bitmapdata);
				image.Dispose();
				return true;
			}
			return false;
		}

		public override void glDraw()
		{
			GL.glClear(GL.GL_COLOR_BUFFER_BIT | GL.GL_DEPTH_BUFFER_BIT);
			GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[0]);

			for (this.loop=0; this.loop < LessonView.num; this.loop++)				// Loop Through All The Stars
			{
				GL.glLoadIdentity();											// Reset The View Before We Draw Each this.star
				GL.glTranslatef(0.0f, 0.0f, this.zoom);							// Zoom Into The Screen (Using The Value In 'zoom')
				GL.glRotatef(this.tilt, 1.0f, 0.0f, 0.0f);						// Tilt The View (Using The Value In 'tilt')
				GL.glRotatef(this.star[this.loop].angle, 0.0f, 1.0f, 0.0f);		// Rotate To The Current this.stars AnGL.gle
				GL.glTranslatef(this.star[this.loop].dist, 0.0f, 0.0f);			// Move Forward On The X Plane
				GL.glRotatef(-this.star[this.loop].angle, 0.0f, 1.0f, 0.0f);	// Cancel The Current this.stars AnGL.gle
				GL.glRotatef(-this.tilt, 1.0f, 0.0f, 0.0f);						// Cancel The Screen Tilt
		
				if (this.twinkle)
				{
					GL.glColor4ub(this.star[(LessonView.num-this.loop)-1].r, this.star[(LessonView.num-this.loop)-1].g, this.star[(LessonView.num-this.loop)-1].b, 255);
					GL.glBegin(GL.GL_QUADS);
					GL.glTexCoord2f(0.0f, 0.0f); GL.glVertex3f(-1.0f,-1.0f, 0.0f);
					GL.glTexCoord2f(1.0f, 0.0f); GL.glVertex3f( 1.0f,-1.0f, 0.0f);
					GL.glTexCoord2f(1.0f, 1.0f); GL.glVertex3f( 1.0f, 1.0f, 0.0f);
					GL.glTexCoord2f(0.0f, 1.0f); GL.glVertex3f(-1.0f, 1.0f, 0.0f);
					GL.glEnd();
				}

				GL.glRotatef(this.spin, 0.0f, 0.0f, 1.0f);
				GL.glColor4ub(this.star[this.loop].r, this.star[this.loop].g, this.star[this.loop].b, 255);
				GL.glBegin(GL.GL_QUADS);
				GL.glTexCoord2f(0.0f, 0.0f); GL.glVertex3f(-1.0f,-1.0f, 0.0f);
				GL.glTexCoord2f(1.0f, 0.0f); GL.glVertex3f( 1.0f,-1.0f, 0.0f);
				GL.glTexCoord2f(1.0f, 1.0f); GL.glVertex3f( 1.0f, 1.0f, 0.0f);
				GL.glTexCoord2f(0.0f, 1.0f); GL.glVertex3f(-1.0f, 1.0f, 0.0f);
				GL.glEnd();

				this.spin += 0.01f;
				this.star[this.loop].angle += (float)(this.loop) / LessonView.num;
				this.star[this.loop].dist -= 0.01f;
				if (this.star[this.loop].dist < 0.0f)
				{
					this.star[this.loop].dist += 5.0f;
					this.star[this.loop].r = (byte)rand.Next(256);
					this.star[this.loop].g = (byte)rand.Next(256);
					this.star[this.loop].b = (byte)rand.Next(256);
				}
			}
		}

		protected override void OnSizeChanged(EventArgs e)
		{
			base.OnSizeChanged(e);
			Size s = Size;

			if (s.Height == 0)
				s.Height = 1;

			GL.glViewport(0, 0, s.Width, s.Height);

			GL.glMatrixMode(GL.GL_PROJECTION);
			GL.glLoadIdentity();
			GL.gluPerspective(45.0f, (double)s.Width /(double) s.Height, 0.1f, 100.0f);	
			GL.glMatrixMode(GL.GL_MODELVIEW);
			GL.glLoadIdentity();
		}

		protected void LessonView_KeyDown(object Sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.Escape)				// Finish the application if the escape key was pressed
				this.finished = true;
			else if (e.KeyCode == Keys.T && !this.tp)
			{
				this.tp = true;
				this.twinkle = !this.twinkle;
			}
			else if (e.KeyCode == Keys.PageUp)			// On page up, move out
				this.zoom -= 0.2f;
			else if (e.KeyCode == Keys.PageDown)		// On page down, move in
				this.zoom += 0.2f;
		}

		private void LessonView_KeyUp(object sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.T)					// Release the twinkle toggle key lock
				this.tp = false;
		}

		protected override bool ProcessDialogKey(Keys keyData)
		{
			if (keyData == Keys.Up)						// Tilt about the x axis
				this.tilt -= 0.5f;
			else if (keyData == Keys.Down)
				this.tilt += 0.5f;
			
			return base.ProcessDialogKey (keyData);
		}
	}
}